### PR TITLE
HW05 Add tests for HW02 and HW03

### PR DESCRIPTION
**Pull request by blooond**

### CODE REVIEW

File: HW05/EgorChernooky/.idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Default ignored files
```
Comment (labs-ruby) [inline HW05/EgorChernooky/.idea/.gitignore:1]: pls don't add your local gitignore to the project, use global gitignore instead to exclude system-specific stuff

File: HW05/EgorChernooky/.idea/EgorChernooky.iml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (labs-ruby) [inline HW05/EgorChernooky/.idea/EgorChernooky.iml:1]: what is it?

File: HW05/EgorChernooky/.idea/misc.xml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (labs-ruby) [inline HW05/EgorChernooky/.idea/misc.xml:1]: pls remove the whole .idea folder

File: HW05/EgorChernooky/array_methods_spec.rb [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+require_relative '/home/egor/homeworks-2021/HW02/EgorChernooky/ArrayMethods'
+
+using ArrayMethods
+
+RSpec.describe ArrayMethods do
+  let(:array) { [1, 2, 3, 4, 5] }
+  let(:mapped_array) { [1, 4, 9, 16, 25] }
+  let(:selected_arr) { [3, 4, 5] }
+  let(:reversed_array) { [5, 4, 3, 2, 1] }
+  let(:empty_array) { [] }
```
Comment (labs-ruby) [inline HW05/EgorChernooky/array_methods_spec.rb:13]: you do not need all of these variables, pls use just one `array` and change it in each context instead

File: HW05/EgorChernooky/task3_spec.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  end
+  let(:numbers) do
+    %w[49.1 32.0]
+  end
+  let(:zero) { 0 }
+
+  describe '#task3' do
+    context 'when there is no logs' do
+      it 'should return 0' do
+        expect(task3(empty_string)).to eq zero
```
Comment (labs-ruby) [inline HW05/EgorChernooky/task3_spec.rb:45]: you do not need `zero` and `empty_string` variables

pls read about predicate matchers

```suggestion
        expect(task3(empty_string)).to be_zero
```

File: HW05/EgorChernooky/array_methods_spec.rb [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+  describe '#my_each' do
+    context 'when array is empty' do
+      it 'should return an empty array' do
+        expect(empty_array.my_each { |item| p item}).to eq empty_array
+      end
+    end
+
+    context 'when array is not empty' do
+      it 'should return an origin array' do
```
Comment (B0RlS) [inline HW05/EgorChernooky/array_methods_spec.rb:23]:
```suggestion
      it 'returns an origin array' do
```

File: HW05/EgorChernooky/array_methods_spec.rb (hunk 4 above)
Comment (B0RlS) [inline HW05/EgorChernooky/array_methods_spec.rb:13]: move lets to contexts
```suggestion
```

File: HW05/EgorChernooky/array_methods_spec.rb [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+RSpec.describe ArrayMethods do
+  let(:array) { [1, 2, 3, 4, 5] }
+  let(:mapped_array) { [1, 4, 9, 16, 25] }
+  let(:selected_arr) { [3, 4, 5] }
+  let(:reversed_array) { [5, 4, 3, 2, 1] }
+  let(:empty_array) { [] }
+
+  describe '#my_each' do
+    context 'when array is empty' do
+      it 'should return an empty array' do
```
Comment (B0RlS) [inline HW05/EgorChernooky/array_methods_spec.rb:17]:
```suggestion
    context 'when array is empty' do
      let(:empty_array) { [] }
      
      it 'should return an empty array' do
```

Comment (blooond) [reply HW05/EgorChernooky/array_methods_spec.rb:14]: All done

Comment (B0RlS) [reply HW05/EgorChernooky/array_methods_spec.rb:14]: @blooond where?

File: HW05/EgorChernooky/array_methods_spec.rb (hunk 6 above)
Comment (B0RlS) [reply HW05/EgorChernooky/array_methods_spec.rb:20]: @blooond

File: HW05/EgorChernooky/array_methods_spec.rb [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        expect(array.my_each { |item| p item }).to eq array
+      end
+    end
+  end
+
+  describe '#my_map' do
+    let(:mapped_array) { [1, 4, 9, 16, 25] }
+
+    context 'when array is empty' do
+      it 'should return an empty array'  do
```
Comment (B0RlS) [inline HW05/EgorChernooky/array_methods_spec.rb:30]:
```suggestion
    context 'when array is empty' do
      let(:empty_array) { [] }
      
      it 'returns an empty array'  do
```

File: HW05/EgorChernooky/array_methods_spec.rb [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    context 'when array is not empty' do
+      it 'should return an origin array' do
+        expect(array.my_each { |item| p item }).to eq array
+      end
+    end
+  end
+
+  describe '#my_map' do
+    let(:mapped_array) { [1, 4, 9, 16, 25] }
```
Comment (B0RlS) [inline HW05/EgorChernooky/array_methods_spec.rb:27]:
```suggestion
```

move this let to context

File: HW05/EgorChernooky/array_methods_spec.rb [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+      it 'should return an empty array'  do
+        expect(empty_array.my_map { |item| item**2 }).to be_empty
+      end
+    end
+
+    context 'when array is not empty' do
+      it 'should return a changed array but origin array should not be changed' do
+        expect(array.my_map { |item| item**2 }).to eq mapped_array
+        expect(array == mapped_array).to eq false
+      end
```
Comment (B0RlS) [inline HW05/EgorChernooky/array_methods_spec.rb:39]:
```suggestion
      let(:mapped_array) { [1, 4, 9, 16, 25] }
      
      it 'returns new array' do
        expect(array.my_map { |item| item**2 }).to eq mapped_array
        expect(array ).not_to eq mapped_array
      end
```

File: HW05/EgorChernooky/array_methods_spec.rb [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    context 'when array is empty' do
+      it 'should return an empty array' do
+        expect(empty_array.my_select { |item| item > 2 }).to be_empty
+      end
+    end
+
+    context 'when array is not empty' do
+      it 'should return a new array and not change the origin one' do
+        expect(array.my_select { |item| item > 2 }).to eq selected_arr
+        expect(array == selected_arr).to eq false
```
Comment (B0RlS) [inline HW05/EgorChernooky/array_methods_spec.rb:72]: same ( line 38 )

File: HW05/EgorChernooky/task1_spec.rb [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    <<~LOGS
+      10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498
+      10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
+      2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
+      10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290
+    LOGS
+  end
+  let(:return_log) do
+    '2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>'
+  end
```
Comment (B0RlS) [inline HW05/EgorChernooky/task1_spec.rb:17]: move it to second context please

File: HW05/EgorChernooky/task1_spec.rb [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+RSpec.describe 'hw_03_t_1' do
+  let(:logs) do
+    <<~LOGS
+      10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0498
+      10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
+      2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
+      10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290
+    LOGS
+  end
+  let(:return_log) do
```
Comment (B0RlS) [inline HW05/EgorChernooky/task1_spec.rb:15]:
```suggestion
  let(:expected_results) do
```

File: HW05/EgorChernooky/task1_spec.rb [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    '2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>'
+  end
+  let(:empty_string) { '' }
+
+  describe '#task1' do
+    context 'when logs are empty' do
+      it 'should return an empty string' do
+        expect(task1(empty_string)).to be_empty
+      end
+    end
```
Comment (B0RlS) [inline HW05/EgorChernooky/task1_spec.rb:25]:
```suggestion
  describe '#task1' do
    context 'when logs are empty' do
      let(:empty_string) { '' }
      
      it 'should return an empty string' do
        expect(task1(empty_string)).to be_empty
      end
    end
```

File: HW05/EgorChernooky/task2_spec.rb [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+  let(:empty_string) { '' }
+  let(:empty_array) { [] }
+  let(:output_array) do
+    [
+      '23/Apr/2018:20:30:39 +0300 FROM: 10.6.246.103 TO: /TEST/2/MESSAGES',
+      '23/Apr/2018:20:30:42 +0300 FROM: 10.6.246.101 TO: /TEST/2/RUN',
+      '23/Apr/2018:20:31:39 +0300 FROM: 10.6.246.101 TO: /TEST/2/MESSAGES'
+    ]
+  end
+  let(:no_suit_var_text) { '2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>' }
```
Comment (B0RlS) [inline HW05/EgorChernooky/task2_spec.rb:24]:
```suggestion
  let(:invalid_log) { '2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>' }
```

Comment (B0RlS) [inline HW05/EgorChernooky/task2_spec.rb:24]: same

File: HW05/EgorChernooky/task4_spec.rb [hunk 16]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+require 'rspec'
+require_relative '/home/egor/homeworks-2021/HW03/EgorChernooky/egor_chernooky_hw_03_t_4'
+
+RSpec.describe 'hw_03_t_4' do
+  let(:str) { 'hel2!lo' }
+  let(:result) { { letters: 5, digits: 1 } }
+
```
Comment (B0RlS) [inline HW05/EgorChernooky/task4_spec.rb:9]: could you cover the case when an empty string is passed to the method please?

File: HW05/EgorChernooky/array_methods_spec.rb [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+    context 'when array is empty' do
+      it 'should return an empty array' do
+        expect(empty_array.my_reverse).to be_empty
+      end
+    end
+
+    context 'when array is not empty' do
+      it 'should return a reversed array without changing the origin one' do
+        expect(array.my_reverse).to eq reversed_array
+        expect(array == reversed_array).to eq false
```
Comment (B0RlS) [inline HW05/EgorChernooky/array_methods_spec.rb:103]: same

**Pull request merged**